vars/yamlSubstituteVariables.groovy

```diff
@@ -0,0 +1,244 @@
+@Field String STEP_NAME = getClass().getName()
+@Field Set GENERAL_CONFIG_KEYS = []
+@Field Set STEP_CONFIG_KEYS = GENERAL_CONFIG_KEYS + [
+    /**
+     * The input Yaml data as `Object`.
+     */
+    'inputYaml',
+    /**
+     * The variables Yaml data as `Object`.
+     * Can be a `List<Map<String, Object>>` or a `Map<String, Object>` and should contain
+     * variables names and values to replace variable references contained in `inputYaml`.
+     */
+    'variablesYaml',
+    /**
+     *  An `com.sap.piper.variablesubstitution.ExecutionContext` that can be used to query
+     *  whether the script actually replaced any variables.
+     */
+    'executionContext'
+]
+
+@Field Set PARAMETER_KEYS = STEP_CONFIG_KEYS
+
+/**
+ * Step that substitutes variables in a given YAML input data object. The format to reference a variable
+ * in YAML data is to use double parentheses `((` and `))`, e.g. `((variableName))`. Variables will be replaced by
+ * values that are read from another Yaml object. The script returns a deep copy of the input YAML with all occurrences
+ * of variables replaced (if they were found in the YAML of variables). The result will be written into the `commonPipelineEnvironment`
+ * by the name `yamlSubstituteVariablesResult` and will be of type `Object`.
+ * <p>
+ * The format follows <a href="https://docs.cloudfoundry.org/devguide/deploy-apps/manifest-attributes.html#variable-substitution">Cloud Foundry standards</a>.
+ * <p>
+ * Usage: `yamlSubstituteVariables inputYaml: <yamlDataObject>, variablesYaml: <yamlDataObject> [, executionContext: context]`
+ *
+ * @param arguments - the map of arguments.
+ */
+@GenerateDocumentation
+void call(Map<String, String> arguments) {
+    handlePipelineStepErrors (stepName: STEP_NAME, stepParameters: arguments) { // synchronous closure call!
+        def script = checkScript(this, arguments)  ?: this
+
+        // load default & individual configuration
+        Map config = ConfigurationHelper.newInstance(this)
+                                        .loadStepDefaults()
+                                        .mixinStepConfig(script.commonPipelineEnvironment, STEP_CONFIG_KEYS)
+                                        .mixinStageConfig(script.commonPipelineEnvironment, arguments.stageName ?: env.STAGE_NAME, STEP_CONFIG_KEYS)
+                                        .mixin(arguments, PARAMETER_KEYS)
+                                        .withMandatoryProperty("inputYaml", "[YamlSubstituteVariables] Input Yaml data must not be null or empty.")
+                                        .withMandatoryProperty("variablesYaml", "[YamlSubstituteVariables] Variables Yaml data must not be null or empty.")
+                                        .use()
+
+        Object inputYaml = config.inputYaml
+        Object variablesYaml = config.variablesYaml
+
+        logger.setConfig(config)
+
+        def result = substitute(inputYaml, variablesYaml, config?.executionContext)
```

Comment:
1.) We strongly rely on the config, so `config.?` (the `.?`)  is IMO inappropriate. If the would be missing it would be better to fail with a NPE. And more: if the config would be missing we would also fail above at `config.?inputYaml`. I would like to suggest to replace `.?` with `.`.

2.) Execution context expects an `com.sap.piper.variablesubstitution.ExecutionContext` from outside. There are now two cases: 'outside'  it another pipeline step. In this case we do not have an issue. if 'outside' is the pipeline itself than it is a problem since we rely in the pipeline on something from `src`. As already mentioned elsewhere we try to keep the API of the piper-lib as small as possible and restricted to the pipeline steps itself (... and not also the scripts/classes inside `src`). But that gets resolved automatically when that gets moved to a util class.

Reply:
Both are changed now.